### PR TITLE
ISPN-7990 Provide a way to remove an entry from all indexes

### PR DESCRIPTION
**Pull request by gustavonalle**
https://issues.jboss.org/browse/ISPN-7990

### CODE REVIEW

**Comment by gustavonalle** [comment]
@karesti Once you have the test to reproduce the issue of lack of values in the context during topology change, you can test against this branch.

**Comment by rvansa** [comment]
@gustavonalle Needs rebase.

**Comment by rvansa** [comment]
LGTM, @anistor could you bless this?

**Comment by anistor** [comment]
Sure, I'll have a look after the meeting.

**Comment by gustavonalle** [comment]
Rebased

File: query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java [hunk 1]
```diff
@@ -232,6 +234,18 @@ public void purgeIndex(Class<?> entityType) {
       purgeIndex(null, entityType);
    }
 
+   /**
+    * Remove entries from all indexes by key
+    */
+   void removeFromIndexes(TransactionContext transactionContext, Object key) {
+      transactionContext = transactionContext == null ? makeTransactionalEventContext() : transactionContext;
```
Comment (anistor) [inline query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java]: Assigning same value to itself is a bit strange. Why not do it like ```if (transactionContext == null) { transactionContext = makeTransactionalEventContext();
}```   ?

File: query/src/main/java/org/infinispan/query/backend/SearchWorkCreator.java [hunk 2]
```diff
@@ -31,4 +31,7 @@
     * @return collection of Work instances
     */
    Collection<Work> createPerEntityWorks(Object entity, Serializable id, WorkType workType);
+
+   Collection<Work> createPerEntityIdWorks(Serializable id, IndexedTypeIdentifier entityType, WorkType workType);
```
Comment (anistor) [inline query/src/main/java/org/infinispan/query/backend/SearchWorkCreator.java]: A bit of javadoc would be nice.

File: query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java (hunk 1 above)
Comment (gustavonalle) [inline query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java]: This used everywhere in the QI

File: query/src/main/java/org/infinispan/query/backend/SearchWorkCreator.java (hunk 2 above)
Comment (gustavonalle) [inline query/src/main/java/org/infinispan/query/backend/SearchWorkCreator.java]: ok. Will change a bit this method, it does not need to return a collection

File: query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java [hunk 3]
```diff
@@ -9,6 +9,8 @@
 import java.util.Set;
```
Comment (gustavonalle) [inline query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java]: This came from the "official" formatter unfortunately. Should we avoid in general to reformat comments?

**Comment by gustavonalle** [comment]
Updated.

File: query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java (hunk 3 above)
Comment (anistor) [inline query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java]: I we don't change any content, then better not reformat them.

File: query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java (hunk 1 above)
Comment (anistor) [inline query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java]: Indeed, and it looks a bit silly. Let's leave like that in the name of tradition :)

**Comment by anistor** [comment]
Integrated in master. Thanks @gustavonalle !

**Pull request merged**